File: golang/generate_parentheses.go
func generateParenthesis(n int) []string {
	result := []string{}
	helper(n, 0, 0, "", &result)
	return result
}

func helper(n int, left int, right int, current string, result *[]string) {
	// base case
	if left == right && left == n {
		*result = append(*result, current)
		return
	}

	// recursion based off left right count
	if left == n {
		helper(n, left, right+1, current+")", result)
	} else if left == right {
		helper(n, left+1, right, current+"(", result)
	} else if left > right {
		helper(n, left+1, right, current+"(", result)
		helper(n, left, right+1, current+")", result)
	}
}

File: typescript/generate_parentheses.ts
function generateParenthesis(n: number): string[] {
	let combos = []
	helper('(', n, 1, 0, combos)
	return combos
}

function helper(current: string, n: number, left: number, right: number, combos: string[]) {
	// base case
	if (current.length == 2 * n) {
		combos.push(current) // assuming current is valid
		return
	}

	if (left == n) {
		helper(current + ')', n, left, right + 1, combos)
		return
	}

	if (left == right) {
		helper(current + '(', n, left + 1, right, combos)
		return
	}

	if (left > right) {
		helper(current + '(', n, left + 1, right, combos)
		helper(current + ')', n, left, right + 1, combos)
		return
	}
}
